dataset/src/main/kotlin/org/jetbrains/kotlinx/dl/dataset/preprocessor/image/ImagePreprocessor.kt

```diff
@@ -11,13 +11,19 @@ import java.awt.image.BufferedImage
 /** Basic interface for image preprocessors. It operates on [BufferedImage]. */
 public interface ImagePreprocessor {
     /**
-     * Transforms [image] with [inputShape] to the new image with the new shape.
+     * Computes output image shape for the provided [inputShape].
+     * @param inputShape image input shape. Null value means that input image size is not known.
+     *                   This is useful for operations with a fixed output size,
+     *                   which should return a non-null value in this case.
      *
-     * @return Pair <new image; new shape>.
+     * @return output image shape
      */
-    public fun apply(image: BufferedImage, inputShape: ImageShape): Pair<BufferedImage, ImageShape>
```

Comment:
So, I had yet two ideas, but didn't implement yet
1) return pair, but calculate outputShape by inputShape for each Preprocessor
2) return new object of class OurGreatImage (with fields BufferedImage and ImageShape), later it's easy to add these prepocessors like extension functions to the OurGreatImage for example

Let's discuss both approaches?

Reply:
Could you explain why a separate shape is needed? `BufferedImage` has this information already, so it seems a bit redundant to me. Not to mention that because of the nullable fields using `ImageShape` might be awkward in this case.